src/compiler/checker.ts

```diff
@@ -21078,24 +21231,49 @@ namespace ts {
             return tryGetTypeAtPosition(signature, pos) || anyType;
         }
 
-        function tryGetTypeAtPosition(signature: Signature, pos: number): Type | undefined {
+        function tryGetAtXPositionWorker<T>(
+            signature: Signature,
+            pos: number,
+            symbolGetter: (symbol: Symbol) => T,
+            restTypeGetter: (type: Type, pos: number, paramCount: number) => T): T | undefined {
             const paramCount = signature.parameters.length - (signature.hasRestParameter ? 1 : 0);
             if (pos < paramCount) {
-                return getTypeOfParameter(signature.parameters[pos]);
+                return symbolGetter(signature.parameters[pos]);
             }
             if (signature.hasRestParameter) {
                 const restType = getTypeOfSymbol(signature.parameters[paramCount]);
-                if (isTupleType(restType)) {
-                    if (pos - paramCount < getLengthOfTupleType(restType)) {
-                        return restType.typeArguments![pos - paramCount];
-                    }
-                    return getRestTypeOfTupleType(restType);
-                }
-                return getIndexTypeOfType(restType, IndexKind.Number);
+                return restTypeGetter(restType, pos, paramCount);
             }
             return undefined;
         }
 
+        function getTupleTypeForArgumentAtPos(restType: Type, pos: number, paramCount: number) {
+            if (isTupleType(restType)) {
+                if (pos - paramCount < getLengthOfTupleType(restType)) {
+                    return restType.typeArguments![pos - paramCount];
+                }
+                return getRestTypeOfTupleType(restType);
+            }
+            return getIndexTypeOfType(restType, IndexKind.Number);
+        }
+
+        function tryGetTypeAtPosition(signature: Signature, pos: number): Type | undefined {
+            return tryGetAtXPositionWorker(signature, pos, getTypeOfParameter, getTupleTypeForArgumentAtPos);
+        }
+
+        function getTupleParameterNameForArgumentAtPos(restType: Type, pos: number, paramCount: number): __String {
+            if (isTupleType(restType)) {
+                if (pos - paramCount < getLengthOfTupleType(restType) && restType.target.associatedNames) {
+                    return restType.target.associatedNames[pos - paramCount];
+                }
+            }
+            return `arg${pos}` as __String;

```

Comment:
We should be consistent with what we already do for synthetic parameter names for tuple spreads.